numpy/core/src/umath/override.c

```diff
@@ -86,367 +79,124 @@ get_array_ufunc_overrides(PyObject *args, PyObject *kwds,
-        if (axes_kwd == NULL && PyErr_Occurred()) {
-            return -1;
-        }
-        if (axis_kwd && axes_kwd) {
-            PyErr_SetString(PyExc_TypeError,
-                            "cannot specify both 'axis' and 'axes'");
+    static PyObject *out_str = NULL;
+    if (out_str == NULL) {
+        out_str = PyUnicode_InternFromString("out");
+        if (out_str == NULL) {
             return -1;
         }
     }
-    /* finally, ufuncs accept 'sig' or 'signature' normalize to 'signature' */
-    return nkwds == 0 ? 0 : normalize_signature_keyword(*normal_kwds);
-}
-
-static int
-normalize_reduce_args(PyUFuncObject *ufunc, PyObject *args,
-                      PyObject **normal_args, PyObject **normal_kwds)
-{
-    /*
-     * ufunc.reduce(a[, axis, dtype, out, keepdims])
-     */
-    npy_intp nargs = PyTuple_GET_SIZE(args);
-    npy_intp i;
-    PyObject *obj;
-    static PyObject *NoValue = NULL;
-    static char *kwlist[] = {"array", "axis", "dtype", "out", "keepdims",
-                             "initial", "where"};
-
-    npy_cache_import("numpy", "_NoValue", &NoValue);
-    if (NoValue == NULL) return -1;
-
-    if (nargs < 1 || nargs > 7) {
-        PyErr_Format(PyExc_TypeError,
-                     "ufunc.reduce() takes from 1 to 7 positional "
-                     "arguments but %"NPY_INTP_FMT" were given", nargs);
-        return -1;
-    }
-    *normal_args = PyTuple_GetSlice(args, 0, 1);
-    if (*normal_args == NULL) {
-        return -1;
-    }
 
-    for (i = 1; i < nargs; i++) {
-        PyObject *kwd = _PyDict_GetItemStringWithError(*normal_kwds, kwlist[i]);
-        if (kwd == NULL && PyErr_Occurred()) {
+    if (out_args != NULL) {
+        /* Replace `out` argument with the normalized version */
+        int res = PyDict_SetItem(normal_kwds, out_str, out_args);
+        if (res < 0) {
             return -1;
         }
-        else if (kwd) {
-            PyErr_Format(PyExc_TypeError,
-                         "argument given by name ('%s') and position "
-                         "(%"NPY_INTP_FMT")", kwlist[i], i);
+    }
+    else {
+        /* Ensure that `out` is not present. */
+        int res = PyDict_Contains(normal_kwds, out_str);
+        if (res < 0) {
             return -1;
         }
-        obj = PyTuple_GET_ITEM(args, i);
-        if (i == 3) {
-            /* remove out=None */
-            if (obj == Py_None) {
-                continue;
-            }
-            obj = PyTuple_GetSlice(args, 3, 4);
-        }
-        /* Remove initial=np._NoValue */
-        if (i == 5 && obj == NoValue) {
-            continue;
-        }
-        PyDict_SetItemString(*normal_kwds, kwlist[i], obj);
-        if (i == 3) {
-            Py_DECREF(obj);
+        if (res) {
+            return PyDict_DelItem(normal_kwds, out_str);
         }
     }
     return 0;
 }
 
+/*
+ * ufunc() and ufunc.outer() accept 'sig' or 'signature';
+ * normalize to 'signature'
+ */
```

Comment:
Maybe add to the comment that by the time this is called it is no longer possible for both to be present.